src/kernels/MIOpenUnfold.cpp

```diff
@@ -0,0 +1,238 @@
+                           L,
+                           LH,
+                           LW,
+                           kernel_size_h,
+                           kernel_size_w,
+                           stride_h,
+                           stride_w,
+                           padding_h,
+                           padding_w,
+                           dilation_h,
+                           dilation_w,
+                           input_tv,
+                           output_tv);
+}
+
+template <typename DTYPE>
+__device__ void unfoldBackward4D(const DTYPE* __restrict__ output_grad,
+                                 DTYPE* __restrict__ input_grad,
+                                 int N,
+                                 int C,
+                                 int H,
+                                 int W,
+                                 int P,
+                                 int L,
+                                 int LH,
+                                 int LW,
+                                 int kernel_size_h,
+                                 int kernel_size_w,
+                                 int stride_h,
+                                 int stride_w,
+                                 int padding_h,
+                                 int padding_w,
+                                 int dilation_h,
+                                 int dilation_w,
+                                 tensor_view_t<3> output_grad_tv,
+                                 tensor_view_t<4> input_grad_tv)
+{
+    /*
+     * output_grad = {N, C * P, L}, input_grad = {N, C, H, W}
+     * where P = kernel_size_h * kernel_size_w, L = # of blocks (see host code for
+     * formula)
+     * => gws = {ceil(N * C * H * W, LOCAL_SIZE)}, lws = {LOCAL_SIZE}
+     */
+
+    const int gid = threadIdx.x + blockIdx.x * blockDim.x;
+    int nch = gid / W, w = gid % W;
+    int nc = nch / H, h = nch % H;
+    int n = nc / C, c = nc % C;
+    if(n >= N)
+        return;
+
+    FLOAT_ACCUM sum = 0.0f;
+    for(int ph = 0; ph < kernel_size_h; ++ph)
+    {
+        for(int pw = 0; pw < kernel_size_w; ++pw)
+        {
+            int lhsh = h - ph * dilation_h + padding_h;
+            int lwsw = w - pw * dilation_w + padding_w;
+            if(lhsh % stride_h != 0)
+                continue;
+            if(lwsw % stride_w != 0)
+                continue;
+            int lh = lhsh / stride_h;
+            int lw = lwsw / stride_w;
+            if(lh < 0 || LH <= lh)
+                continue;
+            if(lw < 0 || LW <= lw)
+                continue;
+            long output_grad_idx = output_grad_tv.stride[2] * (lh * LW + lw) +
+                                   output_grad_tv.stride[1] * (c * P + (ph * kernel_size_w + pw)) +
+                                   output_grad_tv.stride[0] * n;
+            sum += CVT_FLOAT2ACCUM(output_grad[output_grad_idx]);
+        }
+    }
+
+    long input_grad_idx = input_grad_tv.stride[3] * w + input_grad_tv.stride[2] * h +
```

Comment:
Same as above

Reply:
resolve in commit [f3dea16](https://github.com/moreh-dev/MIOpen/pull/37/commits/f3dea16ca398ffa36db95f75d5ef000ad2b74721)